### DISCUSSION TITLE
how to build nicegui to static web html

### DISCUSSION THREAD

**Discussion by bobwatcherx** [zauberzeug/nicegui#2511] [Q&A]
### Question

is there docs how to export to static web html . no fastapi

**Comment by falkoschindler** [comment]
See https://github.com/zauberzeug/nicegui/discussions/313 for a related discussion.

**Reply by bobwatcherx** [reply]
and how to build a static web. For example, I request Api. and the data changes over time and is dynamic. can static web do that . I have tried . but CSS and JS are not found. and in my case it doesn't work if the api data changes

```py
from nicegui import ui
import requests

async def save():
    html = await ui.run_javascript("document.querySelector('html').outerHTML")
    with open("./test.html", "w", encoding="utf-8") as file:
        file.write(html)
    
    
ui.button("Click", on_click=save)
from_api = ui.column()	

def loadapi():
	url = "YOU DYNAMIC API URL HERE"
	response = requests.get(url)
	files = response.json()
	mydata = files['result']['files']

	for x in mydata:
		with from_api:
			with ui.card():
				ui.image(x['single_img'])
				ui.label(x['title'])
				ui.button("add")



loadapi()

ui.run(port=5555)
```

**Reply by falkoschindler** [reply]
Yes, that's one limitation @Diegiwg [mentioned](https://github.com/zauberzeug/nicegui/discussions/313#discussioncomment-4829147):
> it will not save the CSS and JS files, which NiceGUI generates/uses

**Comment by ProjectGreenfieldSolutions** [comment]
@falkoschindler I am considering a project that fetchs images / videos from a remote API. Would the CSS/JS limitation mean that if I had the CSS / JS files needed to operate the static website manually constructed. I could reference those over the ones NiceGUI generates?

**Reply by falkoschindler** [reply]
I'm not sure if I understand correctly. You want to build an app that fetches images and video from a remote API, and distribute or host a serverless HTML version of it? So the fetch must happen in the frontend, I guess.
Sure, you could certainly take the generated HTML file and replace links to JS and CSS dependencies with links to local files. But the NiceGUI app won't be able to connect to a server, so you need to strip this part.
Overall, I don't think it makes too much sense to run a NiceGUI page without a server. While there could be value in dumping a static HTML output, trying to preserve interactivity sounds tedious. Starting from scratch with a plain Vue/Quasar app might be a more promising approach.

**Reply by ProjectGreenfieldSolutions** [reply]
Thank you, yes I was attempting to piggy back off of the original poster. (Should of started from scratch there as well haha)
I suppose your words still ring true though.

For context, my actual goal is to build a website to showcase my ability build, host, and scale up a modern type of website. (e.g. Instagram, Youtube etc...) By fetching from an API I can eliminate the cost/labor of hosting a database. While searching for the potential of eliminating the middleware, Django, I discovered his post. 

In the past, I've used React to produce a /build of the site but I guess I never updated things dynamically based on fetched results. I wonder if that also has the problems that you had brought to my attention with your response.

Suffice to say, I think I'll accept the database win and commence with the nicegui & django setup. I mean having everything into a single language is a huge boon for continued development.
